### PR TITLE
Avoid considering all orders selling fee.

### LINKED ISSUE #75: Could not round solution (nr_exec_orders > max_nr_exec_orders)
The open solver had an issue with rounding a solution for this instance:
https://gnosis-dev-dfusion.s3.amazonaws.com/data/mainnet_dev/open-solver/instances/2020-08-03/instance_5321545_2020-08-03T14%3A10%3A31.561627002%2B00%3A00.json

 File "/app/open_solver/src/core/validation.py", line 41, in validate
   assert nr_exec_orders <= max_nr_exec_orders
AssertionError
(see https://gnosisinc.slack.com/archives/CQZBU9T5J/p1596463904003100)

It happened again in the subsequent batch:
https://gnosis-dev-dfusion.s3.amazonaws.com/data/mainnet_dev/open-solver/instances/2020-08-03/instance_5321546_2020-08-03T14%3A15%3A31.552265513%2B00%3A00.json

### PR DESCRIPTION
**Pull request by marcovc**
Optimizes how the subset of orders selling fee for the buy token imbalance (f_orders) is found.

To compute this subset we compute the xrates fee_token<->buy_token for multiple problems with a set of f_orders that grow iteratively, ordered by decreasing sell/buy limit xrate of the f_orders, and select the subset of f_orders that lead to higher overall objective value. 

The reason we do this, instead of simply going with the `fee_token <-> buy_token` xrate that maximizes the `fee_token <-> buy_token` problem, is because increasing the amount of fee sold for buy token increases the price of buy token, and therefore indirectly affects objective contribution orders on the `buy_token <-> sell_token` pair.

This optimization explores the fact that this indirect effect of growing the subset of f_orders can only monotonically increase the overall objective value. This means the objective value of the problem `fee_token <-> buy_token <-> sell_token` as a function of the xrate between `fee_token <-> buy_token` still has only one optimum.  
 
Therefore while iteratively growing the subset of f_orders, it cannot be the case that the objective goes down and then up.
 
This PR explores this insight to make the code faster.

Test: Added failing instance to e2e tests.

### CODE REVIEW

File: src/token_pair_solver/solver.py [hunk 1]
```diff
@@ -350,8 +350,17 @@ def solve_token_pair_and_fee_token(
...
 
+            # Optimization: Since f_orders are ordered by limit xrate, the objective
+            # as a function of the size of the prefix used has only one optimum.
+            # In other words, we can stop trying to augmenting the set of f_orders
+            # once adding a new f_order degrades the objective.
+            if best_objective is not None and objective < best_objective:
+                break
+
             # Update best solution found so far if necessary.
             if best_objective is None or objective >= best_objective:
```
Comment (twalth3r) [inline src/token_pair_solver/solver.py:365]: Isn't this condition always satisfied after the new preceding if-statement?

**Review by twalth3r — APPROVED** [review]
I don't know what precisely the reason was for the assertion failing in #75 , but would you agree that it could be rather coincidental that the error goes away with this PR? Say that the best objective is only reached in the last iteration and `break` is never called!?

Otherwise, looks good to me!

(just for curiosity - do you have an estimate what the runtime gains are on average?)

File: src/token_pair_solver/solver.py (hunk 1 above)
Comment (marcovc) [reply src/token_pair_solver/solver.py]: Yes, fixed!

**Comment by marcovc** [comment]
> would you agree that it could be rather coincidental that the error goes away with this PR? Say that the best objective is only reached in the last iteration and break is never called!?

Yes that is true. I removed the "closes 75" and will have another go at it in another PR!

**Comment by marcovc** [comment]
> do you have an estimate what the runtime gains are on average

I tested on a few instances I have available and it doesn't seem too significative.

**Pull request merged**